src/install-pnpm/run.ts

```diff
@@ -37,12 +37,78 @@ export async function runSelfInstaller(inputs: Inputs): Promise<number> {
   return exitCode
 }
 
+// Nearly identical to the function `actions/setup-node` uses.
+// See https://github.com/actions/setup-node/blob/39370e3970a6d050c480ffad4ff0ed4d3fdee5af/src/util.ts#L8
+function getPnpmVersionFromFile(versionFilePath: string) {
+  if (!existsSync(versionFilePath)) {
+    throw new Error(
+      `The specified pnpm version file at: ${versionFilePath} does not exist`
+    )
+  }
+
+  const contents = readFileSync(versionFilePath, 'utf8')
+
+  // Try parsing the file as a `package.json` file.
+  try {
+    const manifest = JSON.parse(contents)
```

Comment:
The behavior would be more predictable if you match the extension of `versionFilePath` against `.json` and `.tool-versions` instead of `try`. We can even be stricter: match the basename of `versionFilePath` against `package.json` and `.tool-versions`.